### ISSUE TITLE
Semantic version with underscore portion puts unexpected version in 02packages

### ISSUE THREAD

**Issue by XSven** [briandfoy/cpan-mini-inject#11]
I wanted to "add" a v1.0.0_01 trial release (release status "testing") to my local DarkPAN. `mcpani` reports the wrong version information: For me v1.0.1 is a bugfix version!

```sh
MCPANI_CONFIG=/home/darkpan/.mcpani/config mcpani --verbose --add --module App::runscript --authorid SVW --modversion v1.0.0_01 --file ./App-runscript-v1.0.0_01.tar.gz

Adding File: App-runscript-v1.0.0_01.tar.gz
Author ID: SVW
Module: App::runscript
Version: v1.0.1
To repository: /usr/share/mirrors/darkpan/repository
```

The _modulelist_ file contains the wrong version information too:

```sh
cat /usr/share/mirrors/darkpan/repository/modulelist
App::runscript                   v1.0.1  S/SV/SVW/App-runscript-v1.0.0_01.tar.gz
```

**Comment by briandfoy** [comment]
Noted. Can you send me (or attach) the tarball for that so I can see what you are seeing.

I thought this was an issue that had been reported before, but now I can't find the source. It would have been over 10 years ago anyway.

Have you uploaded something like this to PAUSE and has PAUSE done the right thing? I don't use semantic versioning so I haven't tried this.

I suspect that the issue is actually in one of the modules that extracts versions, and maybe gets confused by semantic versioning since the version object would ignore the `_`, unlike a string version that wouldn't. If that's the case, it might be a problem that CPAN::Mini can't fix.

**Comment by XSven** [comment]
I have now already uploaded two tar balls to PAUSE successfully. The last one has the download link:

https://cpan.metacpan.org/authors/id/S/SV/SVW/App-runscript-v1.0.0_02.tar.gz

I haven't had the use case with a string version like '1.000_001'. At least the EU:MM installer was able to extract the 'v1.0.0_02' version properly. I have used the recommended idiom

`use version 0.9915; our $VERSION = version->declare( 'v1.0.0_02' );`

in the `App::runscript` module.

**Comment by XSven** [comment]
I have tried to debug the the issue. _Makefile.PL_ of `App::runscript` contains

```
  META_MERGE       => {
    'meta-spec' => { version => 2 },
    provides    => {
      'App::runscript' => {
        file    => 'lib/App/runscript.pm',
        version => 'v1.0.0_03',
      },
    },
    ...
  },
```
This leads to a wrong provides version v1.0.3 in the metadata files `(MY)META.json` and `(MY)META.yml`. This wrong information is then used by `CPAN::Mini::Inject`.

**Comment by XSven** [comment]
The problem is this

```
perl -MCPAN::Meta::Converter -E 'say CPAN::Meta::Converter::_clean_version("v1.0.0_03")'
v1.0.3
```

**Comment by briandfoy** [comment]
Thanks for doing the research and feedback. 

This sounds like the bug is something outside of CPAN::Mini::Inject (other than it's using something that doesn't do it correctly). I think the correct place to file the bug is in the [CPAN-Meta repo](https://github.com/Perl-Toolchain-Gang/CPAN-Meta/issues). However, that module hasn't been updated since 2016 and the project appears effectively dead even though there's a lot of conversation about things that could happen. None of that matters if no one is going to make a new release.

This doesn't mean that this module doesn't have work to do. If CPAN::Meta isn't doing things correctly, we have to work around that. That might be a big change though, and I won't be able to get to it soon.

**Comment by briandfoy** [comment]
This is something I hope to be able to look at in the second half of July when I get back to the keyboard.

**Comment by briandfoy** [comment]
Try the code that's at the HEAD of master. I didn't test this and will need to add tests later, but I think this might fix it. You can see where I'm going with it maybe and work with that.

Okay, some of this goes back to the [Lyon Compromise](https://blogs.perl.org/users/brian_d_foy/2016/02/the-lyon-compromise-on-perl-version-formats.html), and I think CPAN::Meta's last release was around that time and may not have been compliant.

CPAN::Mini::Inject uses CPAN::Meta::Converter through Dist::Metadata.

Anyway, with `version`, you need to call `stringify` to get what the user told you, and `normal` to get what some other people thought you should have used. 

    $ perl -Mversion -E 'say version->new(shift)->stringify' v1.0.0_03
    v1.0.0_03

    $ perl -Mversion -E 'say version->new(shift)->normal' v1.0.0_03
    v1.0.3

I think the fix is to monkey patch CPAN::Meta::Converter to use `stringify` instead of `normal`. Although I've reported the issue in https://github.com/Perl-Toolchain-Gang/CPAN-Meta#138, I don't expect any action on it. And, since I think another release of any sort is unlikely, I feel okay overriding some its code.

**Comment by briandfoy** [comment]
I think [RT#110882](https://rt.cpan.org/Public/Bug/Display.html?id=110882) might be the same issue.

**Comment by haarg** [comment]
It's worth noting that `v1.0.0_03` and `v1.0.3` are "equivalent" versions. It likely makes sense for CPAN::Meta and CPAN::Meta::Requirements to use `stringify` rather than `normal`, but trying to treat those two versions as distinct is incorrect.

**Comment by briandfoy** [comment]
version does not think they are equivalent:

    $ perl -Mversion -E 'say version->parse(shift)->is_alpha' v1.0.3

    $ perl -Mversion -E 'say version->parse(shift)->is_alpha' v1.0.0_03
    1

There is a group of people who very much want to believe they are equivalent, but then there is reality. This is why there was a long discussion in [RT#110882](https://rt.cpan.org/Public/Bug/Display.html?id=110882). People saw the problem, didn't know how to fix it, and gave up. They fixed the blocker for version 0.9913 by removing the test it broke in Module::Build, which is really not a fix. People who mark their distributions as alpha versions want them to be alpha versions.

**Comment by haarg** [comment]
```
$ perl -Mversion -E'say version->parse("v1.0.3") == version->parse("v1.0.0_03") ? "equal" : "not equal"'
equal
```

**Comment by haarg** [comment]
As I said, CPAN::Meta and CPAN::Meta::Requirements should likely be using `stringify`. There isn't really any reason not to maintain the original formatting, and it would maintain an underscore through the tooling. And numeric versions already don't get reformatted.

But the numbers are equal and should not be treated as distinct. You should not use `v1.0.0_03` and then later use `v1.0.3` and expect them to be treated as a different version. I wanted to note this because the original post here notes a "v1.0.0_01 trial release" and then a "v1.0.1 bugfix version", implying both existing separately. That will not work correctly.

**Comment by XSven** [comment]
I have mentioned `v1.0.1` only to state that I am **not** approaching to create a stable bug fix release, **but** a trial/release (`v1.0.0_01`) that should later be promoted to the first (initial) release. I have **consistently used everywhere** the `v1.0.0_01` version specification. And later on I have detected that the configure (`perl Makefile.PL`) has already created META files with a bugfix version number specification `v1.0.1` in it.

**Comment by XSven** [comment]
How can two version specifications be equal (equivalent) if one refers to an alpha release and the other one not? 
To me it is a questionable definition of "equality" if the release status (`is_alpha`) wouldn't have been considered.

**Comment by briandfoy** [comment]
The conservation is getting bogged down in the same issues over version that have been problems forever.

CPAN::Mini::Inject is going to use the string the person gave in the command line, and attempt to use the string from the files if it has to guess on its own by looking at the distro. You said you wanted `v1.0.0_003` and that's what you should get because that's what you said you wanted.

I think the tests even reuse the same test distributions but give them different versions.

**Comment by Leont** [comment]
> CPAN::Meta and CPAN::Meta::Requirements should likely be using stringify.

CPAN::Meta::Requirements probably shouldn't: it deals with version ranges not version numbers.

CPAN::Meta probably should, because it is dealing with version numbers in the provide hash.

**Comment by XSven** [comment]
To let you know I have changed my strategy to configure a trial release in EU:MM. I have followed one of the suggestions made by @Grinnz in https://github.com/Perl-Toolchain-Gang/CPAN-Meta/issues/138:

- don't use a leading "v" in dotted decimal version (presumably not a must)
- don't use underscores in dotted decimal version
- set the `release_status` to either "unstable" (means "alpha") or "testing" ( means "beta") in `META_MERGE` in _Makefile.PL_
- set `DISTVNAME` to "Foo-Bar-1.0.0-TRIAL1" in _Makefile.PL_, if this is your 1st trial release

**Comment by Grinnz** [comment]
The leading v is not required if there at least two decimal separators. It is recommended to avoid the case where someone inadvertently releases a version with only one decimal separator, which will then be interpreted differently if there is no leading v.

**Comment by briandfoy** [comment]
The final ruling on this issue:

* first, don't make semantic versions like this, for the reasons people have pointed out
* however, if you do, CPAN::Mini::Inject will take the version string you used and use that. It has an internal patch to CPAN::Meta::Converter, which is the same as Perl-Toolchain-Gang/CPAN-Meta#138
* this does not solve the problem because the META.* files will have the interpreted (numified) version under `provides`. This is completely external to CPAN::Mini::Inject.

**Comment by briandfoy** [comment]
This is released as 1.005. Thanks,
